File: typescript/src/evaluation/grader.ts
/**
 * Out-of-process grader — isolated, tamper-resistant scoring (Feature B).
 *
 * EurekAgent runs its grader in a separate, read-only container so the agent can
 * submit work but never see or modify the scoring logic — that is what stops
 * reward hacking. This module brings the same shape to LoopEngine: a grading
 * contract (GradeResult + is_better) with sentinel scores so an invalid
 * submission can never rank best, an IsolatedGrader that runs grading behind an
 * injected process boundary, and a real subprocess runner.
 */

/** The score and validity of one graded submission. */
export class GradeResult {
  readonly score: number;
  readonly valid: boolean;
  readonly metrics: Record<string, unknown>;

  constructor(opts: { score: number; valid?: boolean; metrics?: Record<string, unknown> }) {
    this.score = opts.score;
    this.valid = opts.valid ?? true;
    this.metrics = opts.metrics ?? {};
  }

  /** Build an invalid result whose sentinel score can never rank best. */
  static invalid(direction: string = 'maximize', metrics: Record<string, unknown> = {}): GradeResult {
    const sentinel = direction === 'maximize' ? -Infinity : Infinity;
    return new GradeResult({ score: sentinel, valid: false, metrics });
  }

  to_dict(): Record<string, unknown> {
    return { score: this.score, valid: this.valid, metrics: { ...this.metrics } };
  }
}

/**
 * Return whether `a` is a strictly better result than `b`. An invalid result is
 * never better; a valid result always beats an invalid one. Among valid results,
 * `direction` decides ('maximize' => higher is better, 'minimize' => lower).
 */
export function is_better(a: GradeResult, b: GradeResult, direction: string = 'maximize'): boolean {
  if (!a.valid) return false;
  if (!b.valid) return true;
  if (direction === 'minimize') return a.score < b.score;
  return a.score > b.score;
}

/**
 * A grading runner: given a serialized submission, return a result mapping (with
 * at least a `score`), or throw. The mapping is serializable — the runner is a
 * process boundary, so live agent objects never cross it.
 */
export type GraderRunner = (submission: Record<string, unknown>) => Promise<Record<string, unknown>>;

/**
 * Grades a submission via an injected (typically out-of-process) runner. The
 * submission is trusted only when the runner returns a finite numeric score: a
 * crash, a missing score, or a non-finite score all collapse to an invalid
 * sentinel result. This is the defense against a tampered or broken grader.
 */
export class IsolatedGrader {
  private readonly _run: GraderRunner;
  private readonly _direction: string;

  constructor(runner: GraderRunner, direction: string = 'maximize') {
    this._run = runner;
    this._direction = direction;
  }

  async grade(submission: Record<string, unknown>): Promise<GradeResult> {
    let raw: Record<string, unknown>;
    try {
      raw = await this._run(submission);
    } catch {
      return GradeResult.invalid(this._direction);
    }

    if (raw === null || typeof raw !== 'object' || !('score' in raw)) {
      return GradeResult.invalid(this._direction);
    }

    const score = (raw as { score: unknown }).score;
    if (typeof score !== 'number' || !Number.isFinite(score)) {
      const metrics = (raw as { metrics?: Record<string, unknown> }).metrics ?? {};
      return GradeResult.invalid(this._direction, metrics);
    }

    return new GradeResult({
      score,
      valid: ((raw as { valid?: unknown }).valid as boolean) ?? true,
      metrics: ((raw as { metrics?: Record<string, unknown> }).metrics) ?? {},
    });
  }
}

/**
 * Build a runner that grades in a genuinely separate process. The grader script
 * reads a JSON submission from stdin and writes a JSON result to stdout. Running
 * it as a child process means the agent's process never touches the grading code.
 */
export function make_subprocess_runner(
  scriptPath: string,
  executable: string = 'bun',
  _timeout: number = 60,
): GraderRunner {
  return async (submission: Record<string, unknown>): Promise<Record<string, unknown>> => {
    const proc = Bun.spawn([executable, scriptPath], {
      stdin: 'pipe',
      stdout: 'pipe',
      stderr: 'pipe',
    });
    proc.stdin.write(JSON.stringify(submission));
    await proc.stdin.end();
    const out = await new Response(proc.stdout).text();
    const err = await new Response(proc.stderr).text();
    const code = await proc.exited;
    if (code !== 0) {
      throw new Error(`Grader exited ${code}: ${err}`);
    }
    return JSON.parse(out) as Record<string, unknown>;
  };
}

File: loopengine/evaluation/grader.py
"""Out-of-process grader — isolated, tamper-resistant scoring (Feature B).

Plain English: EurekAgent runs its grader in a separate, read-only container so
the agent can submit work but never see or modify the scoring logic. That single
idea is what stops "reward hacking" — the agent can't rewrite the judge to give
itself a perfect score.

This module brings the same shape to LoopEngine:

- A grading CONTRACT (`GradeResult` + `is_better`) that mirrors EurekAgent's
  ``grade_submission`` / ``is_better`` pair, including sentinel scores so an
  invalid submission can NEVER rank as best.
- An `IsolatedGrader` that runs the grading behind an injected process boundary.
  If the grading crashes or returns nonsense, the submission is marked invalid
  rather than silently trusted.
- A real `make_subprocess_runner` that grades in a genuinely separate process.
"""

from __future__ import annotations

import asyncio
import json
import math
import sys
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Optional


# ---------------------------------------------------------------------------
# GradeResult — the verdict for one submission
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class GradeResult:
    """The score and validity of one graded submission.

    Attributes:
        score: The numeric score. For invalid submissions this is a sentinel
            (``-inf`` when maximizing, ``+inf`` when minimizing) so they can
            never compare as best.
        valid: Whether the submission was valid and the score is trustworthy.
        metrics: Optional extra measurements (timing, sub-scores, etc.).
    """

    score: float
    valid: bool = True
    metrics: dict[str, Any] = field(default_factory=dict)

    @classmethod
    def invalid(
        cls, direction: str = "maximize", metrics: Optional[dict[str, Any]] = None
    ) -> "GradeResult":
        """Build an invalid result whose sentinel score can never rank best."""
        sentinel = float("-inf") if direction == "maximize" else float("inf")
        return cls(score=sentinel, valid=False, metrics=metrics or {})

    def to_dict(self) -> dict[str, Any]:
        return {"score": self.score, "valid": self.valid, "metrics": dict(self.metrics)}


def is_better(a: GradeResult, b: GradeResult, direction: str = "maximize") -> bool:
    """Return whether ``a`` is a strictly better result than ``b``.

    An invalid result is never better than anything; a valid result always beats
    an invalid one. Among valid results, ``direction`` decides: 'maximize' means
    higher is better, 'minimize' means lower is better.
    """
    if not a.valid:
        return False
    if not b.valid:
        return True
    if direction == "minimize":
        return a.score < b.score
    return a.score > b.score


# A grading runner: given a serialized submission, return a result mapping (with
# at least a ``score``), or raise. The mapping is serializable — the runner is a
# process boundary, so live agent objects never cross it.
GraderRunner = Callable[[dict[str, Any]], Awaitable[dict[str, Any]]]


# ---------------------------------------------------------------------------
# IsolatedGrader — runs grading behind a process boundary
# ---------------------------------------------------------------------------


class IsolatedGrader:
    """Grades a submission via an injected (typically out-of-process) runner.

    The grading logic lives behind ``runner``. Whatever happens there, the
    submission is only trusted when the runner returns a finite numeric score:
    a crash, a missing score, or a non-finite score all collapse to an invalid
    sentinel result. This is the defense against a tampered or broken grader.
    """

    def __init__(
        self,
        runner: GraderRunner,
        direction: str = "maximize",
    ) -> None:
        self._run = runner
        self._direction = direction

    async def grade(self, submission: dict[str, Any]) -> GradeResult:
        try:
            raw = await self._run(submission)
        except Exception:
            # A grader that crashes (or is tampered with) cannot be trusted.
            return GradeResult.invalid(self._direction)

        if not isinstance(raw, dict) or "score" not in raw:
            return GradeResult.invalid(self._direction)

        score = raw.get("score")
        if not isinstance(score, (int, float)) or not math.isfinite(float(score)):
            metrics = raw.get("metrics") if isinstance(raw, dict) else None
            return GradeResult.invalid(self._direction, metrics=metrics)

        return GradeResult(
            score=float(score),
            valid=bool(raw.get("valid", True)),
            metrics=dict(raw.get("metrics", {})),
        )


# ---------------------------------------------------------------------------
# Real out-of-process runner
# ---------------------------------------------------------------------------


def make_subprocess_runner(
    script_path: str,
    python_executable: Optional[str] = None,
    timeout: float = 60,
) -> GraderRunner:
    """Build a runner that grades in a genuinely separate process.

    The grader script reads a JSON submission from stdin and writes a JSON result
    (``{"score": ..., "valid": ..., "metrics": ...}``) to stdout. Running it in a
    child process means the agent's process never touches the grading code.
    """
    exe = python_executable or sys.executable

    async def _runner(submission: dict[str, Any]) -> dict[str, Any]:
        process = await asyncio.create_subprocess_exec(
            exe,
            script_path,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        payload = json.dumps(submission).encode("utf-8")
        out_b, err_b = await asyncio.wait_for(
            process.communicate(payload), timeout=timeout
        )
        if process.returncode != 0:
            detail = err_b.decode("utf-8", "replace")
            raise RuntimeError(f"Grader exited {process.returncode}: {detail}")
        return json.loads(out_b.decode("utf-8"))

    return _runner
